### PR TITLE
Move the method `addScopes` into GIDGoogleUser API

### PR DESCRIPTION
**Pull request by Alex-4-Git**
- add `addScopes` in GIDGoogleUser.h
- move `addScopes' from GIDSignIn.h to GIDSignIn_Private.h

### CODE REVIEW

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 1]
```diff
@@ -44,6 +52,43 @@ typedef void (^GIDUserAuthCompletion)(GIDUserAuth *_Nullable userAuth, NSError *
 // @return NO if there is no user restored from the keychain.
 - (BOOL)restorePreviousSignInNoRefresh;
 
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+
+// Starts an interactive consent flow on iOS to add scopes to the current user's grants.
```
Comment (mdmathias) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:57]: I don't think this comment will show for us. I am also noticing that we use this style for doc comments (e.g., `//` vs `///`) on internal/private APIs, but I wonder why. Don't we want these doc comments to what we expect them to (especially since we have things like `@param` in here)?

File: Samples/Swift/DaysUntilBirthday/Shared/Services/GoogleSignInAuthenticator.swift [hunk 2]
```diff
@@ -105,8 +107,8 @@ final class GoogleSignInAuthenticator: ObservableObject {
       fatalError("No presenting window!")
     }
 
-    GIDSignIn.sharedInstance.addScopes([BirthdayLoader.birthdayReadScope],
-                                       presenting: presentingWindow) { userAuth, error in
+    currentUser?.addScopes([BirthdayLoader.birthdayReadScope],
+                          presenting: presentingWindow) { userAuth, error in
```
Comment (mdmathias) [inline Samples/Swift/DaysUntilBirthday/Shared/Services/GoogleSignInAuthenticator.swift:111]: Indent one space.

File: GoogleSignIn/Sources/GIDSignIn_Private.h (hunk 1 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDSignIn_Private.h:57]: We can discuss this convention in our weekly meetings.

File: Samples/Swift/DaysUntilBirthday/Shared/Services/GoogleSignInAuthenticator.swift (hunk 2 above)
Comment (Alex-4-Git) [reply Samples/Swift/DaysUntilBirthday/Shared/Services/GoogleSignInAuthenticator.swift]: Done.

File: GoogleSignIn/Sources/GIDSignIn_Private.h (hunk 1 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDSignIn_Private.h]: Done.

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 3]
```diff
@@ -14,36 +14,82 @@
...
 
 // Private |GIDSignIn| methods that are used internally in this SDK and other Google SDKs.
 @interface GIDSignIn ()
 
-// Redeclare |currentUser| as readwrite for internal use.
+/// Redeclare |currentUser| as readwrite for internal use.
 @property(nonatomic, readwrite, nullable) GIDGoogleUser *currentUser;
 
-// Private initializer for |GIDSignIn|.
+/// Private initializer for |GIDSignIn|.
```
Comment (mdmathias) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:42]: Worth noting that it's here for testing purposes (so I assume)?

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 4]
```diff
@@ -14,36 +14,82 @@
...
+/// Authenticates with extra options.
 - (void)signInWithOptions:(GIDSignInInternalOptions *)options;
 
-// Restores a previously authenticated user from the keychain synchronously without refreshing
-// the access token or making a userinfo request. The currentUser.profile will be nil unless
-// the profile data can be extracted from the ID token.
-//
-// @return NO if there is no user restored from the keychain.
+/// Restores a previously authenticated user from the keychain synchronously without refreshing
+/// the access token or making a userinfo request. The currentUser.profile will be nil unless
```
Comment (mdmathias) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:49]: Should the second sentence be listed on a new line below the first so that it's in the "Discussion" portion of the documentation (as opposed to the "Summary," which is typically just one sentence).

**Review by mdmathias — APPROVED** [review]
Just two little nits/questions. LGTM.

File: GoogleSignIn/Sources/GIDSignIn_Private.h (hunk 3 above)
Comment (Alex-4-Git) [reply GoogleSignIn/Sources/GIDSignIn_Private.h:42]: No, this is the only init we have for GIDSignIn. It is used by the method `sharedInstance` in GIDSignIn.

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 5]
```diff
@@ -14,36 +14,83 @@
...
+/// the access token or making a userinfo request.
+/// 
+/// The currentUser.profile will be nil unless the profile data can be extracted from the ID token.
+///
+/// @return NO if there is no user restored from the keychain.
 - (BOOL)restorePreviousSignInNoRefresh;
 
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+
+/// Starts an interactive consent flow on iOS to add scopes to the current user's grants.
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:58]: Update the doc comment to reflect the role of this private method as well as its new behavior.

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 6]
```diff
@@ -14,36 +14,83 @@
...
+///     on the main queue.
+- (void)addScopes:(NSArray<NSString *> *)scopes
+    presentingViewController:(UIViewController *)presentingViewController
+                  completion:(nullable void (^)(GIDUserAuth *_Nullable userAuth,
+                                                NSError *_Nullable error))completion
+    NS_EXTENSION_UNAVAILABLE("The add scopes flow is not supported in App Extensions.");
+
+#elif TARGET_OS_OSX
+
+/// Starts an interactive consent flow on macOS to add scopes to the current user's grants
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:77]: Same as above.

File: GoogleSignIn/Sources/GIDGoogleUser.m [hunk 7]
```diff
@@ -181,6 +182,30 @@ - (OIDAuthState *) authState{
...
+                                                NSError *_Nullable error))completion {
+  [GIDSignIn.sharedInstance addScopes:scopes
+             presentingViewController:presentingViewController
+                           completion:completion];
+}
+
+#elif TARGET_OS_OSX || TARGET_OS_MACCATALYST
+
+- (void)addScopes:(NSArray<NSString *> *)scopes
+ presentingWindow:(NSWindow *)presentingWindow
```
Comment (petea) [inline GoogleSignIn/Sources/GIDGoogleUser.m:199]: Check indentation.

File: GoogleSignIn/Sources/GIDGoogleUser.m [hunk 8]
```diff
@@ -181,6 +182,30 @@ - (OIDAuthState *) authState{
   return ((GTMAppAuthFetcherAuthorization *)self.fetcherAuthorizer).authState;
 }
 
+#if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+
+- (void)addScopes:(NSArray<NSString *> *)scopes
```
Comment (petea) [inline GoogleSignIn/Sources/GIDGoogleUser.m:187]: We should have a test case for this method.

File: GoogleSignIn/Sources/GIDGoogleUser.m [hunk 9]
```diff
@@ -181,6 +182,30 @@ - (OIDAuthState *) authState{
...
+    presentingViewController:(UIViewController *)presentingViewController
+                  completion:(nullable void (^)(GIDUserAuth *_Nullable userAuth,
+                                                NSError *_Nullable error))completion {
+  [GIDSignIn.sharedInstance addScopes:scopes
```
Comment (petea) [inline GoogleSignIn/Sources/GIDGoogleUser.m:191]: What happens if this instance of `GIDGoogleUser` isn't the same as `GIDSignIn.sharedInstance.currentUser`?

Comment (petea) [inline GoogleSignIn/Sources/GIDGoogleUser.m:191]: Should the `kGIDSignInErrorCodeNoCurrentUser` error code be used publicly with this change?

File: Samples/Swift/DaysUntilBirthday/Shared/Services/GoogleSignInAuthenticator.swift [hunk 10]
```diff
@@ -83,13 +83,15 @@ final class GoogleSignInAuthenticator: ObservableObject {
   /// - note: Successful requests will update the `authViewModel.state` with a new current user that
   /// has the granted scope.
   func addBirthdayReadScope(completion: @escaping () -> Void) {
+    let currentUser = GIDSignIn.sharedInstance.currentUser
```
Comment (petea) [inline Samples/Swift/DaysUntilBirthday/Shared/Services/GoogleSignInAuthenticator.swift:86]: Use `guard` to unwrap the optional and handle the no `currentUser` case?

**Review by petea — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h [hunk 11]
```diff
@@ -47,8 +47,10 @@ typedef NS_ERROR_ENUM(kGIDSignInErrorDomain, GIDSignInErrorCode) {
   kGIDSignInErrorCodeEMM = -6,
   /// Indicates there is no `currentUser`.
   kGIDSignInErrorCodeNoCurrentUser = -7,
```
Comment (petea) [inline GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h:49]: Is this error still relevant for the public interface?

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h [hunk 12]
```diff
@@ -47,8 +47,10 @@ typedef NS_ERROR_ENUM(kGIDSignInErrorDomain, GIDSignInErrorCode) {
...
+  /// Indicates there is an operation on a previous user.
+  kGIDSignInErrorCodePreviousUser = -8,
```
Comment (petea) [inline GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h:51]: Let's preserve the existing enum values by adding this as `-9`.

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 13]
```diff
@@ -14,36 +14,83 @@
...
+/// the access token or making a userinfo request.
+/// 
+/// The currentUser.profile will be nil unless the profile data can be extracted from the ID token.
+///
+/// @return NO if there is no user restored from the keychain.
 - (BOOL)restorePreviousSignInNoRefresh;
 
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+
+/// Starts an interactive consent flow on iOS to add scopes to the user's grants.
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:58]: Update to reflect that this private method is operating on `currentUser`.

"current user's grants."

File: GoogleSignIn/Sources/GIDSignIn_Private.h [hunk 14]
```diff
@@ -14,36 +14,83 @@
...
+///     on the main queue.
+- (void)addScopes:(NSArray<NSString *> *)scopes
+    presentingViewController:(UIViewController *)presentingViewController
+                  completion:(nullable void (^)(GIDUserAuth *_Nullable userAuth,
+                                                NSError *_Nullable error))completion
+    NS_EXTENSION_UNAVAILABLE("The add scopes flow is not supported in App Extensions.");
+
+#elif TARGET_OS_OSX
+
+/// Starts an interactive consent flow on macOS to add scopes to the user's grants
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignIn_Private.h:77]: Same as above.

File: Samples/ObjC/SignInSample/Source/SignInViewController.m [hunk 15]
```diff
@@ -280,10 +280,11 @@ - (IBAction)disconnect:(id)sender {
 }
 
 - (IBAction)addScopes:(id)sender {
-  [GIDSignIn.sharedInstance addScopes:@[ @"https://www.googleapis.com/auth/user.birthday.read" ]
-             presentingViewController:self
-                           completion:^(GIDUserAuth *_Nullable userAuth,
-                                        NSError *_Nullable error) {
+  GIDGoogleUser *currentUser = GIDSignIn.sharedInstance.currentUser;
+  [currentUser addScopes:@[ @"https://www.googleapis.com/auth/user.birthday.read" ]
+    presentingViewController:self
```
Comment (petea) [inline Samples/ObjC/SignInSample/Source/SignInViewController.m:285]: Check indentation.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 16]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+  NSWindow *presentingWindow = [[NSWindow alloc] init];
+  [currentUser addScopes:@[kNewScope]
+    presentingWindow:presentingWindow
+          completion:nil];
+#endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  
+  [signIn verify];
+}
+
+- (void)testAddScopes_failure_addScopesToPreviousAccount {
```
Comment (petea) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:477]: "PreviousAccount" => "PreviousUser"?

Comment (petea) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:477]: Might be good to add a test for the`currentUser == nil` case.

**Review by petea — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 17]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
   [self waitForExpectationsWithTimeout:1 handler:nil];
 }
 
+- (void)testAddScopes_success {
+  id signIn = OCMClassMock([GIDSignIn class]);
+  OCMStub([signIn sharedInstance]).andReturn(signIn);
+  [[signIn expect] addScopes:OCMOCK_ANY
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+    presentingViewController:OCMOCK_ANY
```
Comment (ch40w31) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:451]: Indent wrapped lines by 4 spaces, see https://google.github.io/styleguide/objcguide.html#method-invocations

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 18]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+  
+  GIDGoogleUser *currentUser = [self googleUserWithAccessTokenExpiresIn:kAccessTokenExpiresIn
+                                                       idTokenExpiresIn:kIDTokenExpiresIn];
+  
+  OCMStub([signIn currentUser]).andReturn(currentUser);
+  
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  UIViewController *presentingViewController = [[UIViewController alloc] init];
+  [currentUser addScopes:@[kNewScope]
+    presentingViewController:presentingViewController
```
Comment (ch40w31) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:465]: Same here regarding indentation for wrapped lines

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 19]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+  GIDGoogleUser *previousUser = [self googleUserWithAccessTokenExpiresIn:kAccessTokenExpiresIn
+                                                        idTokenExpiresIn:kNewIDTokenExpiresIn];
+  
+  XCTestExpectation *expectation =
+      [self expectationWithDescription:@"Completion is called."];
+  
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  UIViewController *presentingViewController = [[UIViewController alloc] init];
+  [previousUser addScopes:@[kNewScope]
+    presentingViewController:presentingViewController
```
Comment (ch40w31) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:495]: Same here regarding indentation for wrapped lines

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 20]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+                  completion:^(GIDUserAuth *userAuth, NSError *error) {
+    [expectation fulfill];
+    XCTAssertNil(userAuth);
+    XCTAssertEqual(error.code, kGIDSignInErrorCodePreviousUser);
+  }];
+#elif TARGET_OS_OSX
+  NSWindow *presentingWindow = [[NSWindow alloc] init];
+  [previousUser addScopes:@[kNewScope]
+         presentingWindow:presentingWindow
+                  completion:^(GIDUserAuth *userAuth, NSError *error) {
```
Comment (ch40w31) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:505]: Align the colons with previous lines

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 21]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
+  UIViewController *presentingViewController = [[UIViewController alloc] init];
+  [currentUser addScopes:@[kNewScope]
+    presentingViewController:presentingViewController
+                  completion:nil];
+#elif TARGET_OS_OSX
+  NSWindow *presentingWindow = [[NSWindow alloc] init];
+  [currentUser addScopes:@[kNewScope]
+    presentingWindow:presentingWindow
+          completion:nil];
```
Comment (ch40w31) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:471]: Align the colons

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 22]
```diff
@@ -437,6 +443,75 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+#elif TARGET_OS_OSX
+  NSWindow *presentingWindow = [[NSWindow alloc] init];
+  [previousUser addScopes:@[kNewScope]
+         presentingWindow:presentingWindow
+                  completion:^(GIDUserAuth *userAuth, NSError *error) {
+    [expectation fulfill];
+    XCTAssertNil(userAuth);
+    XCTAssertEqual(error.code, kGIDSignInErrorCodePreviousUser);
+  }];
+#endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
```
Comment (ch40w31) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:510]: I think the duplication of test code for iOS and macOS can be largely avoided here.

**Review by ch40w31 — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m (hunk 22 above)
Comment (ch40w31) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: Fair point. But if you inline `presentingViewController`/`presentingWindow`, `#if #elif #end` should be localized to one place. Up to you to decide if you want to do this here and other similar places.

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h [hunk 23]
```diff
@@ -45,10 +45,10 @@ typedef NS_ERROR_ENUM(kGIDSignInErrorDomain, GIDSignInErrorCode) {
   kGIDSignInErrorCodeCanceled = -5,
   /// Indicates an Enterprise Mobility Management related error has occurred.
   kGIDSignInErrorCodeEMM = -6,
-  /// Indicates there is no `currentUser`.
-  kGIDSignInErrorCodeNoCurrentUser = -7,
   /// Indicates the requested scopes have already been granted to the `currentUser`.
   kGIDSignInErrorCodeScopesAlreadyGranted = -8,
+  /// Indicates there is an operation on a previous user.
+  kGIDSignInErrorCodePreviousUser = -9,
```
Comment (ch40w31) [inline GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h:51]: It's not clear to me what "previous user" means as an error without reading the comment. I'd suggest a more descriptive name such as "kGIDSignInErrorCodeMismatchWithCurrentUser".

Comment (Alex-4-Git) [reply GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignIn.h]: Done.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m (hunk 22 above)
Comment (petea) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: I think you should limit the `#if ... #endif` to the `presentingViewController` / `presentingWindow` line here as we do elsewhere in the codebase.  You're looking at dropping 30+ lines of repetitious code.

Comment (Alex-4-Git) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: Done.

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m [hunk 24]
```diff
@@ -437,6 +443,95 @@ - (void)testDoWithFreshTokens_handleConcurrentRefresh {
...
+              presentingWindow:OCMOCK_ANY
+#endif // TARGET_OS_IOS || TARGET_OS_MACCATALYST
+                    completion:OCMOCK_ANY];
+  
+  GIDGoogleUser *currentUser = [self googleUserWithAccessTokenExpiresIn:kAccessTokenExpiresIn
+                                                       idTokenExpiresIn:kIDTokenExpiresIn];
+  
+  OCMStub([signIn currentUser]).andReturn(currentUser);
+  
+#if TARGET_OS_IOS || TARGET_OS_MACCATALYST
```
Comment (petea) [inline GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m:462]: Looks like you missed this section.

**Review by petea — APPROVED** [review]

File: GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m (hunk 22 above)
Comment (petea) [reply GoogleSignIn/Tests/Unit/GIDGoogleUserTest.m]: You missed one instance.

File: GoogleSignIn/Sources/GIDGoogleUser.m [hunk 25]
```diff
@@ -181,6 +183,41 @@ - (OIDAuthState *) authState{
...
+  
+#if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+  [GIDSignIn.sharedInstance addScopes:scopes
+             presentingViewController:presentingViewController
+                           completion:completion];
+#elif TARGET_OS_OSX || TARGET_OS_MACCATALYST
+  [GIDSignIn.sharedInstance addScopes:scopes
+                     presentingWindow:presentingWindow
+                           completion:completion];
+#endif // TARGET_OS_IOS && !TARGET_OS_MACCATALYST
```
Comment (ch40w31) [inline GoogleSignIn/Sources/GIDGoogleUser.m:218]: Similarly, apply `#if #elif #endif` only to the `presentingViewController`/`presentingWindow` line to minimize duplication.

**Review by ch40w31 — APPROVED** [review]
